### PR TITLE
test/e2e: add aardvark specific tests

### PR DESCRIPTION
**Pull request by lsm5**
Signed-off-by: Lokesh Mandvekar <lsm5@fedoraproject.org>

### CODE REVIEW

**Comment by lsm5** [comment]
@baude can you check the `Docker-py Compat` test. Nothing in the PR should affect the `alpine` image from what I can see.

File: test/e2e/run_aardvark_test.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+		ctrIP := podmanTest.Podman([]string{"inspect", "--format", fmt.Sprintf(`"{{.NetworkSettings.Networks.%s.IPAddress}}"`, netName), cid})
+		ctrIP.WaitWithDefaultTimeout()
+		cip := ctrIP.OutputToString()
+		Expect(cip).To(MatchRegexp(IPRegex))
+		Expect(ip).Should(Exit(0))
+
+		reverseLookup := podmanTest.Podman([]string{"exec", cid, "dig", "+short", "-x", cip})
+		reverseLookup.WaitWithDefaultTimeout()
+		Expect(reverseLookup).Should(Exit(0))
+		revListArray := reverseLookup.OutputToStringArray()
```
Comment (lsm5) [inline test/e2e/run_aardvark_test.go:66]: @baude can you PTAL here? In manual tests, the `exec $CID dig ...` command shows me the name and id entries, but for whatever reason the Array seems to be empty here.

**Comment by lsm5** [comment]
Tests 4 and 5 pass atm.

**Comment by baude** [comment]
Does this help?

```
diff --git a/test/e2e/run_aardvark_test.go b/test/e2e/run_aardvark_test.go
index f00c92aa4..0e125713d 100644
--- a/test/e2e/run_aardvark_test.go
+++ b/test/e2e/run_aardvark_test.go
@@ -39,7 +39,7 @@ var _ = Describe("Podman run networking", func() {
 
 	})
 
-	It("Aardvark Test 1: One container", func() {
+	FIt("Aardvark Test 1: One container", func() {
 		netName := createNetworkName("Test")
 		session := podmanTest.Podman([]string{"network", "create", netName})
 		session.WaitWithDefaultTimeout()
@@ -49,14 +49,13 @@ var _ = Describe("Podman run networking", func() {
 		ctrID := podmanTest.Podman([]string{"run", "-dt", "--name", "aone", "--network", netName, nginx})
 		ctrID.WaitWithDefaultTimeout()
 		cid := ctrID.OutputToString()
-		ip := podmanTest.Podman([]string{"exec", cid, "dig", "+short", "google.com"})
+		ip := podmanTest.Podman([]string{"exec", cid, "dig", "+short", "aone"})
 		ip.WaitWithDefaultTimeout()
 		Expect(ip.OutputToString()).To(MatchRegexp(IPRegex))
 		Expect(ip).Should(Exit(0))
-
 		ctrIP := podmanTest.Podman([]string{"inspect", "--format", fmt.Sprintf(`"{{.NetworkSettings.Networks.%s.IPAddress}}"`, netName), cid})
 		ctrIP.WaitWithDefaultTimeout()
-		cip := ctrIP.OutputToString()
+		cip := strings.ReplaceAll(ctrIP.OutputToString(), "\"", "")
 		Expect(cip).To(MatchRegexp(IPRegex))
 		Expect(ip).Should(Exit(0))
 
@@ -66,7 +65,7 @@ var _ = Describe("Podman run networking", func() {
 		revListArray := reverseLookup.OutputToStringArray()
 		Expect(revListArray).Should(HaveLen(2))
 		Expect(strings.TrimRight(revListArray[0], ".")).To(Equal("aone"))
-		Expect(strings.TrimRight(revListArray[1], ".")).To(Equal(cid[12]))
+		Expect(strings.TrimRight(revListArray[1], ".")).To(Equal(cid[:12]))
 
 	})
 
@@ -124,7 +123,7 @@ var _ = Describe("Podman run networking", func() {
 		})
 	*/
 
-	FIt("Aardvark Test 4: Two containers, different subnets", func() {
+	It("Aardvark Test 4: Two containers, different subnets", func() {
 		netNameA := createNetworkName("TestA")
 		sessionA := podmanTest.Podman([]string{"network", "create", netNameA})
 		sessionA.WaitWithDefaultTimeout()
@@ -168,7 +167,7 @@ var _ = Describe("Podman run networking", func() {
 		Expect(resB1A1.OutputToString()).To(Equal(""))
 	})
 
-	FIt("Aardvark Test 5: Two containers on their own subnets, one container on both", func() {
+	It("Aardvark Test 5: Two containers on their own subnets, one container on both", func() {
 		netNameA := createNetworkName("TestA")
 		sessionA := podmanTest.Podman([]string{"network", "create", netNameA})
 		sessionA.WaitWithDefaultTimeout()
```

**Comment by lsm5** [comment]
Test 2 flakes like so:

```
 Failure [3.752 seconds]
Podman run networking
/home/lsm5/go/src/github.com/containers/podman/test/e2e/run_aardvark_test.go:14
  Aardvark Test 2: Two containers, same subnet [It]
  /home/lsm5/go/src/github.com/containers/podman/test/e2e/run_aardvark_test.go:73

  Expected
      <[]string | len:0, cap:0>: nil
  to have length 2

  /home/lsm5/go/src/github.com/containers/podman/test/e2e/run_aardvark_test.go:128
```

**Comment by lsm5** [comment]
Test 1 flakes similarly:
``` 
Failure [2.869 seconds]
Podman run networking
/home/lsm5/go/src/github.com/containers/podman/test/e2e/run_aardvark_test.go:14
  Aardvark Test 1: One container [It]
  /home/lsm5/go/src/github.com/containers/podman/test/e2e/run_aardvark_test.go:42

  Expected
      <[]string | len:0, cap:0>: nil
  to have length 2

  /home/lsm5/go/src/github.com/containers/podman/test/e2e/run_aardvark_test.go:67
```

**Comment by lsm5** [comment]
https://github.com/containers/podman/pull/13238/checks?check_run_id=5224084457 is failing. Looks unrelated.

**Review by flouthoc — APPROVED** [review]
@lsm5 LGTM. 

Should we fix flakes before merging, I guess adding delay of `0.5` between running container and trying to dig should fix flakes as of now.

File: test/e2e/run_aardvark_test.go [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+
+		ctrID := podmanTest.Podman([]string{"run", "-dt", "--name", "aone", "--network", netName, nginx})
+		ctrID.WaitWithDefaultTimeout()
+		cid := ctrID.OutputToString()
+		ip := podmanTest.Podman([]string{"exec", cid, "dig", "+short", "aone"})
+		ip.WaitWithDefaultTimeout()
+		Expect(ip.OutputToString()).To(MatchRegexp(IPRegex))
+		Expect(ip).Should(Exit(0))
+
+		ctrIP := podmanTest.Podman([]string{"inspect", "--format", fmt.Sprintf(`"{{.NetworkSettings.Networks.%s.IPAddress}}"`, netName), cid})
```
Comment (Luap99) [inline test/e2e/run_aardvark_test.go:56]:
```suggestion
		ctrIP := podmanTest.Podman([]string{"inspect", "--format", fmt.Sprintf(`{{.NetworkSettings.Networks.%s.IPAddress}}`, netName), cid})
```
Remove the extra quotes here then you do not have to remove then two lines below, same for the other tests where you do this.

**Review by Luap99 — COMMENTED** [review]
If the tests are already flaking on this PR we have to fix it before merging.

File: test/e2e/run_aardvark_test.go (hunk 2 above)
Comment (lsm5) [reply test/e2e/run_aardvark_test.go:56]: thanks, that makes things simpler. I've made that change locally, will check for flakes and push in a bit.

**Comment by lsm5** [comment]
added some timeouts and ran the current version across 20 nodes with 20 flakeAttempts, and 0 flakes so far.

PTAL all.

**Comment by baude** [comment]
> added some sleeps and ran the current version across 20 nodes with 20 flakeAttempts, and 0 flakes so far.
> 
> PTAL all.

Im not super pleased to use timeouts long-term.  I'll look at where you used them and see if I can find a root cause.

@Luap99 if you have ideas, i'd love to hear them.

**Comment by Luap99** [comment]
I think with dns we always need some special timeouts or have flakes, netavark doesn't wait for aardvark to update so there is not much we can do. Since this is just a SIGHUB I don't see us fixing this without changing the way netavark calls aardvark,

**Comment by lsm5** [comment]
if we only need to care about the case of 3 nodes and 3 flakeAttempts, I can probably reduce the sleeps by quite a bit and investigate further.

**Comment by flouthoc** [comment]
If we don't want sleep one generic way could be to create a function in `podman` which verifies committed aardvark entries and the perform dns operation. Cause `sighup` is almost immediate as soon as we write entries.

**Comment by TomSweeneyRedHat** [comment]
LGTM and happy green test buttons

File: test/e2e/run_aardvark_test.go [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+		ctr2.WaitWithDefaultTimeout()
+		Expect(ctr2).Should(Exit(0))
+
+		ctrIP2 := podmanTest.Podman([]string{"inspect", "--format", fmt.Sprintf(`{{.NetworkSettings.Networks.%s.IPAddress}}`, netName), "atwo"})
+		ctrIP2.WaitWithDefaultTimeout()
+		Expect(ctrIP2).Should(Exit(0))
+		cip2 := ctrIP2.OutputToString()
+		Expect(cip2).To(MatchRegexp(IPRegex))
+
+		time.Sleep(5 * time.Second)
```
Comment (Luap99) [inline test/e2e/run_aardvark_test.go:167]: need to remove the sleep

File: test/e2e/run_aardvark_test.go [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+		Expect(ctr2).Should(Exit(0))
+		cid2 := ctr2.OutputToString()
+
+		ctrIP2 := podmanTest.Podman([]string{"inspect", "--format", fmt.Sprintf(`{{.NetworkSettings.Networks.%s.IPAddress}}`, netName), cid2})
+		ctrIP2.WaitWithDefaultTimeout()
+		Expect(ctrIP2).Should(Exit(0))
+		cip2 := ctrIP2.OutputToString()
+		Expect(cip2).To(MatchRegexp(IPRegex))
+
+		time.Sleep(4 * time.Second)
```
Comment (Luap99) [inline test/e2e/run_aardvark_test.go:106]: same

File: test/e2e/run_aardvark_test.go [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+		session.WaitWithDefaultTimeout()
+		defer podmanTest.removeNetwork(netName)
+		Expect(session).Should(Exit(0))
+
+		ctrID := podmanTest.Podman([]string{"run", "-dt", "--name", "aone", "--network", netName, nginx})
+		ctrID.WaitWithDefaultTimeout()
+		Expect(ctrID).Should(Exit(0))
+		cid := ctrID.OutputToString()
+
+		time.Sleep(4 * time.Second)
```
Comment (Luap99) [inline test/e2e/run_aardvark_test.go:54]: same

**Review by Luap99 — CHANGES_REQUESTED** [review]

**Comment by lsm5** [comment]
@baude @Luap99 @flouthoc PTAL. still flakes on rare runs though.

**Comment by flouthoc** [comment]
@lsm5 LGTM. If hope flakes are extremely rare, we can increase retry limit by a few number if this becomes problem.

**Review by flouthoc — APPROVED** [review]

**Pull request merged**